larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,503 @@
+/**
+ *  @file   larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.cc
+ *
+ *  @brief  Implementation of the hit width cluster merging algorithm class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArTwoDReco/HitWidthClusterMergingAlgorithm.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArGeometryHelper.h"
+#include "larpandoracontent/LArObjects/LArTwoDSlidingFitResult.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+
+HitWidthClusterMergingAlgorithm::HitWidthClusterMergingAlgorithm() :
+  m_clusterListName(),
+  m_maxConstituentHitWidth(0.5),
+  m_hitWidthScalingFactor(1.0),
+  m_useSlidingLinearFit(false),
+  m_layerFitHalfWindow(20),
+  m_fittingWeight(10),
+  m_minClusterWeight(0.5),      
+  m_maxXMergeDistance(5),     
+  m_maxZMergeDistance(2),     
+  m_maxMergeCosOpeningAngle(0.97),
+  m_maxDirectionDeviationCosAngle(0.90)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void HitWidthClusterMergingAlgorithm::GetListOfCleanClusters(const ClusterList *const pClusterList, ClusterVector &clusterVector) const
```

Comment:
so, this algorithm inherits from the ClusterAssociations base class algorithm which requires you to define three functions: GetListOfCleanClusters, PopulateClusterAssociationMap, and IsExtremal. The base calls the first function to get a list of relevant clusters. It will then call the second to determine forward and backward associations from which it will then investigate the pathways of associations. During this process it uses the IsExtremal function to work out how to move forward and backward along the associations (i should have mentioned this in the PR comments - sorry)